core/src/main/java/org/everit/json/schema/CombinedSchema.java

```diff
@@ -174,8 +175,20 @@ public ValidationCriterion getCriterion() {
         return criterion;
     }
 
+    // ensure subschemas of type CombinedSchema are always visited first
     public Collection<Schema> getSubschemas() {
-        return subschemas;
+
+        Collection<Schema> combined = subschemas.stream()
+                .filter(schema -> (schema instanceof CombinedSchema))
+                .collect(Collectors.toList());
+
+        Collection<Schema> others = subschemas.stream()
+                .filter(schema -> !(schema instanceof CombinedSchema))
+                .collect(Collectors.toList());
+
+        combined.addAll(others);
```

Comment:
As far as I understand here you add the `others` to a list that was returned by `Collectors.toList()`. The problem with that is, there is no guarantee for the mutability of this list. As [the javadoc says](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Collectors.html#toList()):

> There are no guarantees on the type, mutability, serializability, or thread-safety of the List returned;

So please replace this method implementation with any of the followings:
 - a plain old `for` loop and collecting into two separate ArrayLists
 - `Collectors#toCollection(Supplier)` so that you can make sure your stream collects into a mutable list
 -  or you can have only one stream and sort it based on a comparator sorting by an `instanceof CombinedSchema` expression

Reply:
Updated the PR with requested changes and added more test coverage. Took the comparator approach.